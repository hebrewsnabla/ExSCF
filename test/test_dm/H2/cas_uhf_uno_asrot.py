from pyscf import gto, scf
from fch2py import fch2py
from ortho import check_orthonormal
from pyscf import lib
import numpy as np
import os
from py2fch import py2fch
from uno import uno
from construct_vir import construct_vir
from lo import pm
from assoc_rot import assoc_rot
from shutil import copyfile

lib.num_threads(1)

mol = gto.M()
#   2 atom(s)
mol.atom = '''
H1              0.00000000        0.00000000        0.00000000
H2              0.00000000        0.00000000        2.00000001
'''

mol.basis = {
'H1': gto.basis.parse('''
H     S
    0.544717800E+01    0.156284979E+00
    0.824547240E+00    0.904690877E+00
H     S
    0.183191580E+00    0.100000000E+01
'''),
'H2': gto.basis.parse('''
H     S
    0.544717800E+01    0.156284979E+00
    0.824547240E+00    0.904690877E+00
H     S
    0.183191580E+00    0.100000000E+01
''')}

# Remember to check the charge and spin
mol.charge = 0
mol.spin = 0
mol.verbose = 4
mol.build()

mf = scf.UHF(mol)
mf.max_cycle = 1
mf.max_memory = 4000 # MB
mf.kernel()

# read MOs from .fch(k) file
nbf = mf.mo_coeff[0].shape[0]
nif = mf.mo_coeff[0].shape[1]
S = mol.intor_symmetric('int1e_ovlp')
Sdiag = S.diagonal()
alpha_coeff = fch2py('cas_uhf.fch', nbf, nif, Sdiag, 'a')
beta_coeff = fch2py('cas_uhf.fch', nbf, nif, Sdiag, 'b')
mf.mo_coeff = (alpha_coeff, beta_coeff)
# read done

# check if input MOs are orthonormal
check_orthonormal(nbf, nif, mf.mo_coeff[0], S)
check_orthonormal(nbf, nif, mf.mo_coeff[1], S)

dm = mf.make_rdm1()
mf.max_cycle = 10
mf.kernel(dm)

# transform UHF canonical orbitals to UNO
na = np.sum(mf.mo_occ[0]==1)
nb = np.sum(mf.mo_occ[1]==1)
idx, noon, alpha_coeff = uno(nbf, nif, na, nb, mf.mo_coeff[0], mf.mo_coeff[1], S)
alpha_coeff = construct_vir(nbf, nif, idx[1], alpha_coeff, S)
mf.mo_coeff = (alpha_coeff, beta_coeff)
# done transform

# save the UNO into .fch file
os.system('fch_u2r cas_uhf.fch')
os.rename('cas_uhf_r.fch', 'cas_uhf_uno.fch')
py2fch('cas_uhf_uno.fch', nbf, nif, mf.mo_coeff[0], Sdiag, 'a', noon)
# save done

# associated rotation
npair = np.int64((idx[1]-idx[0]-idx[2])/2)
idx2 = idx[0] + npair - 1
idx3 = idx2 + idx[2]
idx1 = idx2 - npair
idx4 = idx3 + npair
occ_idx = range(idx1,idx2)
vir_idx = range(idx3,idx4)
print(idx1, idx2, idx3, idx4)
occ_loc_orb = pm(mol.nbas, mol._bas[:,0], mol._bas[:,1],  mol._bas[:,3], mol.cart, nbf, npair, mf.mo_coeff[0][:,occ_idx], S, 'mulliken')
vir_loc_orb = assoc_rot(nbf, npair, mf.mo_coeff[0][:,occ_idx], occ_loc_orb, mf.mo_coeff[0][:,vir_idx])
mf.mo_coeff[0][:,occ_idx] = occ_loc_orb.copy()
mf.mo_coeff[0][:,vir_idx] = vir_loc_orb.copy()
# localization done

# save associated rotation MOs into .fch(k) file
copyfile('cas_uhf_uno.fch', 'cas_uhf_uno_asrot.fch')
noon = np.zeros(nif)
py2fch('cas_uhf_uno_asrot.fch', nbf, nif, mf.mo_coeff[0], Sdiag, 'a', noon)
