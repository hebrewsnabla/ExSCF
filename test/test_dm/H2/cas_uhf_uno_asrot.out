

******** <class 'pyscf.scf.uhf.UHF'> ********
method = UHF
initial guess = minao
damping factor = 0
level_shift factor = 0
DIIS = <class 'pyscf.scf.diis.CDIIS'>
diis_start_cycle = 1
diis_space = 8
SCF conv_tol = 1e-09
SCF conv_tol_grad = None
SCF max_cycles = 1
direct_scf = True
direct_scf_tol = 1e-13
chkfile to save SCF result = /share/home/srwang/ExSCF/test/test_dm/tmpo33prgrl
max_memory 4000 MB (current use 65 MB)
number electrons alpha = 1  beta = 1
Set gradient conv threshold to 3.16228e-05
init E= -0.64480089472713
  alpha nocc = 1  HOMO = -0.248574002023546  LUMO = -0.0900508482769706
  beta  nocc = 1  HOMO = -0.241945729803242  LUMO = -0.0916003517709199
cycle= 1 E= -0.907637828086771  delta_E= -0.263  |g|= 0.0146  |ddm|= 0.599
SCF not converged.
SCF energy = -0.907637828086771 after 1 cycles  <S^2> = 6.5933669e-07  2S+1 = 1.0000013

Orthonormality check: j=   2 i=   4 maxv= 5.78868792E-09

Orthonormality check: j=   2 i=   4 maxv= 5.78868790E-09


******** <class 'pyscf.scf.uhf.UHF'> ********
method = UHF
initial guess = minao
damping factor = 0
level_shift factor = 0
DIIS = <class 'pyscf.scf.diis.CDIIS'>
diis_start_cycle = 1
diis_space = 8
SCF conv_tol = 1e-09
SCF conv_tol_grad = None
SCF max_cycles = 10
direct_scf = True
direct_scf_tol = 1e-13
chkfile to save SCF result = /share/home/srwang/ExSCF/test/test_dm/tmpo33prgrl
max_memory 4000 MB (current use 72 MB)
number electrons alpha = 1  beta = 1
Set gradient conv threshold to 3.16228e-05
init E= -0.997011505982131
  alpha nocc = 1  HOMO = -0.494330108557827  LUMO = 0.120177229582684
  beta  nocc = 1  HOMO = -0.494330108557827  LUMO = 0.120177229582684
cycle= 1 E= -0.997011505945357  delta_E= 3.68e-11  |g|= 1.97e-10  |ddm|= 8.53e-10
  alpha nocc = 1  HOMO = -0.494330108581292  LUMO = 0.120177229475714
  beta  nocc = 1  HOMO = -0.494330108581293  LUMO = 0.120177229475714
Extra cycle  E= -0.997011505945358  delta_E= -4.44e-16  |g|= 5.97e-11  |ddm|= 2.46e-10
converged SCF energy = -0.997011505945358  <S^2> = 0.91048261  2S+1 = 2.154514

The orthonormality of Alpha MO after PAO construction:
maxv=    0.0000000000
abs_mean=    0.0000000000

PM orbital localization begins: using mulliken population
Warning in subroutine pm: only 1 orbital. No rotation.
0 1 1 2
